Issue/cif 876

deepkamal opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description

Core/WCM Teaser component extended to support Product or category page linking.

## Related Issue

CIF-876

## Motivation and Context

The Teaser component is part of the AEM Core components library. Currently, this component allows authors to attach image, title, description and CTA (to link to site content sections). We need to extend this component so that an author can link the Teaser component to a product or category.

 Extend AEM Core Component - Teaser (http://opensource.adobe.com/aem-core-wcm-components/library/teaser.html) to link Call to Action button to product/category
 Authors will be able to use the product/category picker to select the product/category to link the Call to Action button.

## How Has This Been Tested?

Unit test cases, manual testing

## Screenshots (if appropriate):

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes and the overall coverage did not decrease.
- [x] All unit tests pass on CircleCi.
- [x] I ran all tests locally and they pass.

deepkamal commented on bundles/core/src/main/java/com/adobe/cq/commerce/core/components/internal/models/v1/teaser/CifTeaserImpl.java:
```diff
@@ -0,0 +1,185 @@
+@Model(adaptables = SlingHttpServletRequest.class, adapters = CifTeaser.class, resourceType = CifTeaserImpl.RESOURCE_TYPE)
+public class CifTeaserImpl implements CifTeaser {
+
+    protected static final String RESOURCE_TYPE = "core/wcm/components/commerce/teaser/v1/teaser";
```
fixed this, it was oversight

herzog31 commented on bundles/core/src/main/java/com/adobe/cq/commerce/core/components/internal/models/v1/teaser/CifTeaserImpl.java:
```diff
@@ -0,0 +1,185 @@
+        }
+    }
+
+    void testHelper(Resource mockedResource, Page productPage, Page categoryPage) {
```
@deepkamal please have a look at the other tests. There we use `Whitebox` provided by `org.mockito.internal.util.reflection.Whitebox`, not the `powermock` one.

deepkamal replied:
yes, fixed this.

deepkamal commented on bundles/core/src/main/java/com/adobe/cq/commerce/core/components/internal/models/v1/teaser/CifTeaserImpl.java:
```diff
@@ -0,0 +1,185 @@
+        Resource actionsNode = resource.getChild(CifTeaser.NN_ACTIONS);
+        if (actionsNode != null) {
+            for (Resource action : actionsNode.getChildren()) {
+                actions.add(new ListItem() {
```
this is how it was done in original sling model, (`https://github.com/adobe/aem-core-wcm-components/blob/master/bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/TeaserImpl.java#L189`) , should modify it to separate out a class, pl suggest.

herzog31 replied:
I let you decide what you like better. I prefer having a separate class, so it can be tested outside of the context of the sling model, which makes testing easier.
Please make sure that in your unit tests you cover all cases of the class, especially `getURL`.

deepkamal replied:
`getURL` is actually tested in `CifTeaserImplTest`.
I will separate `ListItem` implementation in a class.

deepkamal replied:
new class created `com.adobe.cq.commerce.core.components.internal.models.v1.teaser.CommerceTeaserActionItem`

deepkamal commented on bundles/core/pom.xml:
```diff
@@ -494,6 +501,7 @@
         <dependency>
             <groupId>com.adobe.cq</groupId>
             <artifactId>core.wcm.components.core</artifactId>
+            <version>2.0.0</version>
```
fixed.

mhaack commented on bundles/core/src/main/java/com/adobe/cq/commerce/core/components/models/teaser/CifTeaser.java:
```diff
@@ -0,0 +1,105 @@
+import com.adobe.cq.wcm.core.components.models.ListItem;
+
+@ProviderType
+public interface CifTeaser {
```
Why do we need our own interface here at all? Why not extend or use `com.adobe.cq.wcm.core.components.models.Teaser`?

deepkamal replied:
we need to define `PN_ACTION_PRODUCT_SKU` and `PN_ACTION_CATEGORY_ID` as defined `com.adobe.cq.wcm.core.components.models.Teaser` for `link`. if these two definitions can be included in `CifTeaserImpl` , we can directly implement the wcm Teaser interface - pl suggest.

mhaack replied:
Why not extend `com.adobe.cq.wcm.core.components.models.Teaser` the and only add `PN_ACTION_PRODUCT_SKU` and `PN_ACTION_CATEGORY_ID`?

herzog31 replied:
Relevant read here: https://github.com/adobe/aem-core-wcm-components/wiki/Delegation-Pattern-for-Sling-Models

deepkamal replied:
> Why not extend `com.adobe.cq.wcm.core.components.models.Teaser` the and only add `PN_ACTION_PRODUCT_SKU` and `PN_ACTION_CATEGORY_ID`?
yeah, I'll do this now,

deepkamal replied:
CommerceTeaser interface now extends core wcm Teaser interface

mhaack commented on bundles/core/src/main/java/com/adobe/cq/commerce/core/components/models/teaser/CifTeaser.java:
```diff
@@ -0,0 +1,105 @@
+    /**
+     * Name of the resource property that defines whether or not the teaser has Call-to-Action elements
+     *
+     * @since com.adobe.cq.wcm.core.components.models 12.4.0
```
These comments are all copy and paste from sites core component.

deepkamal replied:
mostly yes, but for `PN_ACTION_CATEGORY_ID` and ` PN_ACTION_PRODUCT_SKU` the verbose is changed to describe the property.

deepkamal replied:
CommerceTeaser interface now extends core wcm Teaser interface

deepkamal commented on README.md:
```diff
@@ -24,6 +24,7 @@ See our [wiki](https://github.com/adobe/aem-core-wcm-components/wiki) for usage
 - [Navigation](ui.apps/src/main/content/jcr_root/apps/core/cif/components/structure/navigation/v1/navigation)
 - [Search Results](ui.apps/src/main/content/jcr_root/apps/core/cif/components/commerce/searchresults/v1/searchresults)
 - [Shopping Cart](ui.apps/src/main/content/jcr_root/apps/core/cif/components/commerce/minicart/v1/minicart)
+- [CIF Teaser](ui.apps/src/main/content/jcr_root/apps/core/cif/components/content/teaser/v1/teaser)
```
`Commerce Teaser` suits better, I will change if there are no other suggestions.

deepkamal replied:
renamed top CommerceTeaser, readme updated

deepkamal commented on bundles/core/src/main/java/com/adobe/cq/commerce/core/components/internal/models/v1/teaser/CifTeaserImpl.java:
```diff
@@ -0,0 +1,172 @@
+import com.fasterxml.jackson.annotation.JsonIgnore;
+
+@Model(adaptables = SlingHttpServletRequest.class, adapters = CifTeaser.class, resourceType = CifTeaserImpl.RESOURCE_TYPE)
+public class CifTeaserImpl implements CifTeaser {
```
are you suggesting to keep class names as `TeaserImpl` and Interface `Teaser`? I can refactor.

mhaack replied:
If we really need (see comment below) I would stay with just `Teaser`

deepkamal replied:
I was thinking, since this model is used for action nodes, we can rather call it `TeaserAction`, or `CommerceTeaserAction` in general `<OurNameForComponent>Action` , shall we?

deepkamal replied:
renamed to CommerceTeaser , CommerceTeaserImpl
CommerceTeaser interface now extends core wcm Teaser interface

deepkamal commented:
> Honestly, I don't get what this is so complex and has some many files. If we base this on the sites teaser component, why do we need to overlay the HTL files at all? Why do we not implement the sites teaser interfaces?

`ui.apps/src/main/content/jcr_root/apps/core/cif/components/content/teaser/v1/teaser/action.html` can be removed, and based on what we decide about `CifTeaser` interface, we can reduce that as well, rest all files are required, `CifTeaserImpl` can be reduced depending upon the review discussion.

mhaack commented:
@deepkamal looks like something went wrong with that last merge/rebase. The PR has now many unrelated files and even worse your changes are messed up as well. There is for example `CommerceTeaser` now which looks good ;-) but there is still `CifTeaser`?

deepkamal commented:
> @deepkamal looks like something went wrong with that last merge/rebase. The PR has now many unrelated files and even worse your changes are messed up as well. There is for example `CommerceTeaser` now which looks good ;-) but there is still `CifTeaser`?

Indeed, sorry about the mess up, there was some sync done between master and my branch on github web, 
and then I made wrong decision to pull and merge from master on my local repo after making changes in local files, later when i attempted push it conflicted with GH hosted code,and after pull the deleted(actually renamed) files appeared again and in haste i missed it, 
long story short - I have removed the obsolete files, and I see changes in plenty of files coming from Master merge, thus now I have reverted to previous commit just before the master merge, and pushed changes with `--force`

deepkamal commented:
I have synched this with master (earlier issues fixed) and worked on all noted comments, its ready for review.

mhaack commented on bundles/core/pom.xml:
```diff
@@ -177,7 +178,8 @@
                         <artifactId>maven-bundle-plugin</artifactId>
                         <configuration>
                             <instructions>
-                                <Bundle-RequiredExecutionEnvironment>JavaSE-1.${aem.java.version}</Bundle-RequiredExecutionEnvironment>
```
Can this be reverted?

deepkamal replied:
done.

mhaack commented on ui.apps/src/main/content/jcr_root/apps/core/cif/components/content/teaser/v1/teaser/teaser.html:
```diff
@@ -0,0 +1,32 @@
+  ~ See the License for the specific language governing permissions and
+  ~ limitations under the License.
+  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/-->
+<div data-sly-use.teaser="com.adobe.cq.wcm.core.components.models.Teaser"
```
Since `CommerceTeaser` extends teaser do we actually need both models here? Would it be easier to only use `data-sly-use.teaser="com.adobe.cq.commerce.core.components.models.teaser.CommerceTeaser"`

deepkamal replied:
`CommerceTeaser` only models the actions, thus we need to keep both `Teaser` and `CommerceTeaser` as well, the sightly code uses `CommerceTeaser` to render actions.

mhaack commented:
This looks much cleaner now.

herzog31 commented on bundles/core/pom.xml:
```diff
@@ -366,7 +368,12 @@
+        <dependency>
+            <groupId>org.jetbrains</groupId>
+            <artifactId>annotations</artifactId>
+            <version>RELEASE</version>
```
Please use a fixed version here. Please also check https://github.com/adobe/aem-core-cif-components/pull/89#discussion_r304249641 and see if `javax.annotation` provides an equal annotation.

deepkamal replied:
thanks, using javax.annotation.Nonnull instead, dep removed.

herzog31 commented on bundles/core/src/test/java/com/adobe/cq/commerce/core/components/internal/models/v1/teaser/CommerceTeaserImplTest.java:
```diff
@@ -0,0 +1,122 @@
+        List<ListItem> actionItems = this.slingModel.getActions();
+
+        Assert.assertTrue(this.slingModel.isActionsEnabled());
+        Assert.assertTrue(actionItems.size() == 4);
```
You add 4 actions in the test, but you use asserts only for 3 of them. What about the 4th case?

deepkamal replied:
here, the behavior is that if niether productSKU nor CategoryID is defined, then Link will point to `currentPage` , skipped this in testing as we are not setting `currentPage` in slingModel mocks in the test class - do you think it should also be asserted?

herzog31 replied:
If you don't assert it, why do you add it here https://github.com/adobe/aem-core-cif-components/pull/74/files/b22fa76672e0b9773b0a52a966bfa832f65a5831#diff-cb87fd5300d137fc2056c8ad9a1a7224R63 then?

deepkamal replied:
Did so to verify that even null ProductSKU and null CategoryId with a title is also getting added in action arraylist verified through `Assert.assertTrue(actionItems.size() == 4)` , I will add a mocked currentPage as well, just to cover 4rth assertion too. Thanks.

deepkamal replied:
well what do you know, the added test case to realize its failing - fixed it in actionItem class.
this is done.

herzog31 replied:
Good job 👍
